### ISSUE TITLE
Dev WF Actionable PRs: Milestone 1 Remainders

### ISSUE THREAD

**Issue by garath** [dotnet/arcade#7097]
Tracking issue to help coordinate the Milestone 1 effort.

Goal: Process a build from end-to-end "build complete" to "checks tab has text and the run is closed"

Tasks:

| Done | Item | Assignee | Expected Completion Date |
| -- | -- | -- | -- |
| | Merge PR 13509 Failure Guessing / Renderer Integration | | | 
| | &nbsp;&nbsp;&nbsp;&nbsp;Unit tests | | |
| &check; | &nbsp;&nbsp;&nbsp;&nbsp;Renderer -> check suite handoff | | |
| &check; | &nbsp;&nbsp;&nbsp;&nbsp;Fix Timeline data in Failure Guessing | | |
| &check; | &nbsp;&nbsp;&nbsp;&nbsp;Fix build versus test leg | | |
| &check; | Merge PR 13542 Renderer / Check Suite integration | Ali | 3/18 | 
| | &nbsp;&nbsp;&nbsp;&nbsp;Unit tests | | |
| &check; | DI for TestManagementHttpClient and BuildHttpClient within the Helix infrastructure | | | 
| &check; | PAT with vso.build and vso.test_write for these clients | | | 
| | Testing for Failure Guessing | | | 
| &check; | Ability to deploy Build Result Analysis background processor in Helix Service Fabric cluster | Missy | |

Low priority, candidate to bump to next milestone:

- [ ] Update AzDO Build Complete system hook to use "Storage Queue hook" pointing to new storage account queue
- [ ] Update queue consumer in Failure Guessing to consume this form
- [x] Be able to handle multiple build pipelines for the same PR - Chad

**Comment by garath** [comment]
I'm switching to FR today. Chad has created a branch devwf-m1 that picks up PR 13509.

From what I see the remaining items are:

- Add the authenticated AzDO Clients to the DI service collection. I believe the consumption in the Guesser is all set up. 
- Some work to allow testing of the Guessing and Renderer without AzDO. From discussions, it sounds like this is best done by a custom HttpMessageHandler that can be inserted into the *HttpClients.

**Comment by missymessa** [comment]
@garath are your remaining items reflected in the checkboxes in the issue details, or do we need to add them?

**Comment by garath** [comment]
> are your remaining items reflected in the checkboxes in the issue details, or do we need to add them?

They are reflected in the checkboxes.

**Comment by ChadNedzlek** [comment]
The checkbox for the "timeline" stuff is checked, but I can't find the code that does anything with it. Nothing seems to be setting the "BuildStepResult" value of the analysis.

**Comment by garath** [comment]
> The checkbox for the "timeline" stuff is checked, but I can't find the code that does anything with it. Nothing seems to be setting the "BuildStepResult" value of the analysis.

Ah, you're right, it doesn't fill in BuildStepResult. It does get Timeline data for determining if the build was a retry in BuildWasRetried().

**Comment by ChadNedzlek** [comment]
We also need a checkbox for handling more than one pipeline. We can't turn this on for runtime until that's true, otherwise the builds are going to stomp all over each other pretty unpredictably.

**Comment by missymessa** [comment]
> We also need a checkbox for handling more than one pipeline. We can't turn this on for runtime until that's true, otherwise the builds are going to stomp all over each other pretty unpredictably.

Added this to the checkboxes.

**Comment by ChadNedzlek** [comment]
I did a little cleanup and got all the services registered so the dependency validation tests pass. The two BIG things left are multiple pipelines and the timeline/build error stuff.

The multiple pipelines is going to require changes in the render code, because it currently treats a build as the top level thing, so we'll need to change that.

**Comment by markwilkie** [comment]
I'm OK cutting multiple timeline stuff for M1

**Comment by missymessa** [comment]
Verified that the BuildResultAnalysis Processor was deployed to the SF cluster on staging, so we should not have any surprises when we deploy this in production.

**Comment by ChadNedzlek** [comment]
@markwilkie I don't think it's cuttable.  If we don't do it, I don't think we can realistically turn it on for any real customers, because as each pipeline comes in, it will obliterate the previous answer.  I don't think it's going to be _pretty_ or _efficient_, but we should just be able to glue all the builds together and form a sort of concatenation of all the answers.

**Comment by markwilkie** [comment]
I thought runtime was the only one w/ multiple pipelines?  (maybe I'm being silly)

**Comment by ChadNedzlek** [comment]
aspnetcore also does (they have the "real" pipeline and the "quarantine" pipeline, both of which take almost exactly 83 minutes to complete).  We could turn it on for arcade and maybe like, sdk or wpf?  Is that "done" enough for us?

**Comment by missymessa** [comment]
Validation: 

- [ ] Turn on GitHub App in Arcade Services
- [ ] Validate the Build Result Analysis functionality
  - [ ] Make a PR that includes breaking tests
  - [ ] Make a PR that includes a breaking build (and does not have breaking tests)
  - [ ] Validate both PRs have the BuildResultAnalysis Summary (a.k.a. Checks Tab)
  - [ ] Validate that the failed tests appear on the summary for the PR with the failed tests
  - [ ] Validate that the failed build results appear on the summary for the PR with the failed builds
  - [ ] Validate that successful PRs are not blocked by the BuildResultAnalysis Summary (if it appears) 
- [ ] Validate the sentiment tracker

**Comment by missymessa** [comment]
Sentiment tracker validation: based off feedback in our channel, it appears to be working, however, the GitHub is blocking the use of our tracking image. This won't block us from collecting sentiment data, but it will prevent us from having a more accurate count of how many times the summary check page is viewed.

**Comment by garath** [comment]
> the GitHub is blocking the use of our tracking image. This won't block us from collecting sentiment data, but it will prevent us from having a more accurate count of how many times the summary check page is viewed.

We were concerned about this at first, but we later determined that GitHub's proxy behavior does not prevent the accurate count.

**Comment by missymessa** [comment]
Ali created this PR: https://github.com/dotnet/arcade-services/pull/1559/checks?check_run_id=2152487413 to validate:  

```
Turn it on for the correct repo first (arcade-services), validate it works
Intentionally turn it on for some other, wrong repo that's alst not really customer facing, to ensure the code correctly does nothing
Turn it on for everything and let the validation code handle it
```

It uncovered a bug that we later fixed. @AlitzelMendez or @ChadNedzlek, can you comment on this issue what the bug was when you get a chance?

**Comment by missymessa** [comment]
Ali's PR also validates breaking tests.

**Comment by missymessa** [comment]
Validation for a failing build: https://github.com/dotnet/arcade-services/pull/1561/checks?check_run_id=2188373273

**Comment by missymessa** [comment]
Watched for other's PRs in Arcade services (https://github.com/dotnet/arcade-services/pull/1563) and verified that successful PRs are not blocked by the Build Result Analysis check (because it doesn't need to exist).

**Comment by missymessa** [comment]
Closing this issue. Outstanding tests and the low priority stuff will be moved into separate issues.
